src/server/scripts/Northrend/Naxxramas/boss_kelthuzad.cpp

```diff
@@ -268,355 +144,218 @@ class boss_kelthuzad : public CreatureScript
 
     struct boss_kelthuzadAI : public BossAI
     {
-        boss_kelthuzadAI(Creature* creature) : BossAI(creature, BOSS_KELTHUZAD), spawns(creature)
-        {
-            Initialize();
-            uiFaction = me->getFaction();
-        }
-
-        void Initialize()
-        {
-            nGuardiansOfIcecrownCount = 0;
-
-            nAbomination = 0;
-            nWeaver = 0;
-        }
-
-        uint32 uiFaction;
-
-        uint8  nGuardiansOfIcecrownCount;
-        uint8  nAbomination;
-        uint8  nWeaver;
-
-        std::map<ObjectGuid, float> chained;
-
-        SummonList spawns; // adds spawn by the trigger. kept in separated list (i.e. not in summons)
-
-        void ResetPlayerScale()
-        {
-            std::map<ObjectGuid, float>::const_iterator itr;
-            for (itr = chained.begin(); itr != chained.end(); ++itr)
+        public:
+            boss_kelthuzadAI(Creature* creature) : BossAI(creature, BOSS_KELTHUZAD), _frostboltCooldown(0), _phaseThree(false), _guardianCount(0)
             {
-                if (Player* charmed = ObjectAccessor::GetPlayer(*me, itr->first))
-                    charmed->SetObjectScale(itr->second);
+                for (uint8 i = 0; i < nGuardianSpawns; ++i)
+                    _guardianGroups[i] = SUMMON_GROUP_GUARDIAN_FIRST + i;
             }
 
-            chained.clear();
-        }
-
-        void Reset() override
-        {
-            _Reset();
-
-            me->setFaction(35);
-            me->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NON_ATTACKABLE | UNIT_FLAG_DISABLE_MOVE | UNIT_FLAG_NOT_SELECTABLE);
-
-            ResetPlayerScale();
-            spawns.DespawnAll();
-
-            instance->SetData(DATA_ABOMINATION_KILLED, 0);
-
-            if (GameObject* trigger = ObjectAccessor::GetGameObject(*me, instance->GetGuidData(DATA_KELTHUZAD_TRIGGER)))
+            void InitializeAI() override { Reset(); }
+            void JustRespawned() override
             {
-                trigger->ResetDoorOrButton();
-                trigger->SetPhaseMask(1, true);
+                instance->SetBossState(BOSS_KELTHUZAD, NOT_STARTED);
```

Comment:
you don't need InitializeAI()/JustRespawned() calling for Reset(), it will be called without them, and the boss state is already set in _Reset();

The need for a  void JustRespawned() override { } would be only to avoid Reset() being called twice(which wouldn't matter in this case)